### PR TITLE
refactor to guzzle client and environment variables

### PR DESCRIPTION
**Pull request by DevKCode**

### CODE REVIEW

File: src/Forms/TurnstileCaptchaField.php [hunk 1]
```diff
@@ -172,71 +175,57 @@ public function validate($validator)
...
-        if (!function_exists('curl_init')) {
-            user_error('You must enable php-curl to use this field', E_USER_ERROR);
-            return false;
-        }
-
-        $secret_key = $this->_secretKey ?: self::config()->secret_key;
-        $url = 'https://challenges.cloudflare.com/turnstile/v0/siteverify';
-        $ch = curl_init($url);
+        $curlOptions = [];
+        $secret_key = $this->_secretKey ?: Environment::getEnv('SS_TURNSTILE_SECRET_KEY');
```
Comment (blueo) [inline src/Forms/TurnstileCaptchaField.php]: please put these into a getSecretKey function

File: src/Forms/TurnstileCaptchaField.php [hunk 2]
```diff
@@ -172,71 +175,57 @@ public function validate($validator)
...
 
-        if (!function_exists('curl_init')) {
-            user_error('You must enable php-curl to use this field', E_USER_ERROR);
-            return false;
-        }
-
-        $secret_key = $this->_secretKey ?: self::config()->secret_key;
-        $url = 'https://challenges.cloudflare.com/turnstile/v0/siteverify';
-        $ch = curl_init($url);
+        $curlOptions = [];
```
Comment (blueo) [inline src/Forms/TurnstileCaptchaField.php]: these curl options (Proxy/ssl) can all be removed now we have the httpclient class

File: src/Forms/TurnstileCaptchaField.php [hunk 3]
```diff
@@ -172,71 +175,57 @@ public function validate($validator)
...
-                $validator->validationError($this->name, _t(
-                    'Terraformers\\TurnstileCaptcha\\Forms\\CaptchaField.EMPTY',
-                    '_Please answer the captcha,
-                     if you do not see the captcha you must enable JavaScript'),
-                    'validation');
-                return false;
-            }
-
-        } else {
+        $curlOptions[CURLOPT_RETURNTRANSFER] = true;
```
Comment (blueo) [inline src/Forms/TurnstileCaptchaField.php]: remove all the curl options

File: src/Forms/TurnstileCaptchaField.php [hunk 4]
```diff
@@ -172,71 +175,57 @@ public function validate($validator)
...
+           $response = $client->request('POST', 'https://challenges.cloudflare.com/turnstile/v0/siteverify',[
+               'json' => [
+                   'secret' => $secret_key,
+                   'response' => $recaptchaResponse,
+                   'remoteip' => $request->getIP(),
+               ],
+               'curl' => $curlOptions,
+               'timeout' =>10
+           ]);
+       }catch (Throwable $e) {
```
Comment (blueo) [inline src/Forms/TurnstileCaptchaField.php]: [nitpick] can we run the phpcs linter on this to tidy up spacing etc?

File: src/Forms/TurnstileCaptchaField.php [hunk 5]
```diff
@@ -172,71 +175,57 @@ public function validate($validator)
...
+           if($e instanceof RequestException && $e->hasResponse()){
```
Comment (blueo) [inline src/Forms/TurnstileCaptchaField.php]: this is swallowing quite a few exceptions - instead I think we should add a general validation error if something happens here and log the message using the logger interface (see line 226). So remove this if check, add a validation error here (probably the same as on line 223) and log the error - then return false. (see https://docs.guzzlephp.org/en/stable/quickstart.html#exceptions for the exceptions this is missing)

File: src/Forms/TurnstileCaptchaField.php [hunk 6]
```diff
@@ -172,71 +175,57 @@ public function validate($validator)
...
+        if ($response->getStatusCode() !== 200) {
             $validator->validationError($this->name, _t(
-                'Terraformers\\TurnstileCaptcha\\Forms\\CaptchaField.VALIDATE_ERROR',
+                'Terraformers\\TurnstileCaptcha\\Forms\\TurnstileCaptchaField.VALIDATE_ERROR',
                 '_Captcha could not be validated'),
                 'validation');
             $logger = Injector::inst()->get(LoggerInterface::class);
             $logger->error(
-                'Captcha validation failed as request was not successful.'
+                'Turnstile Captch Field validation failed as request was not successful.'
```
Comment (blueo) [inline src/Forms/TurnstileCaptchaField.php]: typo

File: src/Forms/TurnstileCaptchaField.php [hunk 7]
```diff
@@ -288,48 +277,48 @@ public function getCaptchaTheme(): string
...
     public function setSecretKey($key)
     {
         $this->_secretKey = $key;
     }
 
     /**
-     * Setter for _proxyServer to allow injector config to override the value
+     * Setter for _proxyServer, this will override the injector or environment variable configuration
      */
     public function setProxyServer($server)
```
Comment (blueo) [inline src/Forms/TurnstileCaptchaField.php]: these proxy and the verify ssl methods can be removed

File: src/Http/HttpClient.php [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+namespace Terraformers\TurnstileCaptcha\Http;
+
+use GuzzleHttp\Client;
+use SilverStripe\Core\Injector\Injectable;
+
+class HttpClient
+{
+  use Injectable;
+
+  protected ?Client $client;
```
Comment (blueo) [inline src/Http/HttpClient.php]: spacing here

File: src/Forms/TurnstileCaptchaField.php [hunk 9]
```diff
@@ -172,71 +175,57 @@ public function validate($validator)
...
-                     if you do not see the captcha you must enable JavaScript'),
-                    'validation');
-                return false;
-            }
-
-        } else {
+        $curlOptions[CURLOPT_RETURNTRANSFER] = true;
+        $curlOptions[CURLOPT_SSL_VERIFYPEER] = self::config()->verify_ssl;
+        $curlOptions[CURLOPT_USERAGENT] = str_replace(',', '/', 'SilverStripe');
+        $client = HttpClient::create()->getClient();
```
Comment (blueo) [inline src/Forms/TurnstileCaptchaField.php]: can you add the client class as a dependency instead - eg https://docs.silverstripe.org/en/5/developer_guides/extending/injector/#dependencies-and-properties as this makes it easier to change with config

File: tests/TestTurnstileCaptchaField.php [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * Testing the TurnstileCaptchaField Spam Protection Field
+     * @return void
+     */
+    public function testSpamProtectionField(): void
+    {
+        $form = Form::create(Controller::create(), 'Form', new FieldList(), new FieldList());
+
+        $turnstileCaptchField = new TurnstileCaptchaField('turnstileField');
+        $turnstileCaptchField->setSiteKey(Environment::getEnv('SS_TURNSTILE_SITE_KEY'));
```
Comment (blueo) [inline tests/TestTurnstileCaptchaField.php]: a test probably shouln't use env vars - you can use the testing keys or fake keys

File: tests/TestTurnstileCaptchaField.php [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+                }')
+        ]);
+        $handleStack = HandlerStack::create($mock);
+        $client = new Client(['handler' => $handleStack]);
+        // mock response data
+        $response = $client->send($request);
+        $responseData = json_decode($response->getBody(), true);
+        // validating the response data
+        $this->assertTrue($responseData['success']);
+        $this->assertEmpty($responseData['error-codes']);
```
Comment (blueo) [inline tests/TestTurnstileCaptchaField.php]: can we add a test in here ensuring that the validation will pass with a successful response and fail on an unsuccessful one?

File: src/Forms/TurnstileCaptchaField.php [hunk 12]
```diff
@@ -141,10 +144,10 @@ public function __construct($name, $title = null, $value = null)
     public function Field($properties = array())
     {
         $siteKey = $this->getSiteKey();
-        $secretKey = $this->_secretKey ? $this->_secretKey : self::config()->secret_key;
+        $secretKey = $this->_secretKey ? $this->_secretKey :  Environment::getEnv('SS_TURNSTILE_SECRET_KEY');
```
Comment (blueo) [inline src/Forms/TurnstileCaptchaField.php]: this should use a getSecretKey function

File: src/Forms/TurnstileCaptchaField.php [hunk 13]
```diff
@@ -2,50 +2,53 @@
...
-     * @config NocaptchaField.secret_key
+     * @config TurnstileCaptchaField.secret_key
      */
     private static ?string $secret_key = null;
 
 
     /**
      * CURL Proxy Server location
-     * @config NocaptchaField.proxy_server
+     * @config TurnstileCaptchaField.proxy_server
```
Comment (blueo) [inline src/Forms/TurnstileCaptchaField.php]: these properties (Proxy/ssl) can be removed as the client class is where that should happen

File: src/Forms/TurnstileCaptchaField.php [hunk 14]
```diff
@@ -288,48 +277,48 @@ public function getCaptchaTheme(): string
...
 
     /**
-     * Gets the site key configured via NocaptchaField.site_key this is used in the template
+     * Gets the site key configured via TurnstileCaptchaField.site_key this is used in the template
      * @return string
      */
     public function getSiteKey(): string
     {
-        return $this->_sitekey ? $this->_sitekey : self::config()->site_key;
+        return $this->_sitekey ? $this->_sitekey : Environment::getEnv('SS_TURNSTILE_SITE_KEY');
```
Comment (blueo) [inline src/Forms/TurnstileCaptchaField.php]: this should also have a config option (eg keep self::config()->site_key)

**Review by blueo — CHANGES_REQUESTED** [review]
nice update Dev, just a few more bits to tidy up now we're using guzzle and env vars

File: src/Forms/TurnstileCaptchaField.php [hunk 15]
```diff
@@ -172,71 +175,57 @@ public function validate($validator)
...
+               'curl' => $curlOptions,
+               'timeout' =>10
+           ]);
+       }catch (Throwable $e) {
+           if($e instanceof RequestException && $e->hasResponse()){
+               return  $e->getResponse();
+           }
+       }
+
+        if ($response->getStatusCode() !== 200) {
```
Comment (blueo) [inline src/Forms/TurnstileCaptchaField.php]: the docs say to check the 'outcome' property of the response so I wouldn't rely on just the response code. They could use that to indicate you successfully submitted a response but with a failed 'outcome'

**Review by blueo — CHANGES_REQUESTED** [review]

File: src/Forms/TurnstileCaptchaField.php (hunk 15 above)
Comment (blueo) [reply src/Forms/TurnstileCaptchaField.php]: see https://developers.cloudflare.com/turnstile/get-started/server-side-validation/

Comment (blueo) [reply src/Forms/TurnstileCaptchaField.php]: eg if I decode the body of the response when I use an invalid key I get:

image

File: README.md [hunk 16]
```diff
@@ -31,8 +31,8 @@ $form->enableSpamProtection();
 
 ## Configuration
 There are multiple configuration options for the field, you must set the
-site_key and the secret_key which you can get from the [reCAPTCHA
-page](https://www.google.com/recaptcha). These configuration options must be
+site_key and the secret_key which you can get from the [turnstile
+page](https://developers.cloudflare.com/turnstile/). These configuration options must be
 added to your site's yaml config typically this is mysite/\_config/config.yml.
 ```yml
 Terraformers\TurnstileCaptcha\Forms\TurnstileCaptchaField:
```
[the remark this answers is not in the document]
Comment (DevKCode) [reply README.md:43]: I added like this, let me know If this works 
`SilverStripe\Core\Injector\Injector:
    Terraformers\TurnstileCaptcha\Http\HttpClient:
        class: App\HttpClient`

**Review by blueo — APPROVED** [review]

**Pull request merged**